Improve Kedro-Viz docs for Experiment Tracking

tynandebold opened the pull request:
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.

## Description

Resolves (part of) #2144.

> **Warning**
> This PR shouldn't be merged until version 5.2.0 of Kedro-Viz is released. This is because there's content in this PR that will only exist when that release is out.

## Development notes

Taken from the comment [here](https://github.com/kedro-org/kedro/issues/2144#issuecomment-1375703174), I've done the following:

- [x] Ensure that the spaceflights tutorial works without issue. There was a problem with Plotly where the charts weren't rendering (relates to https://github.com/kedro-org/kedro-viz/issues/1205).
- [x] There are actually two pages for exp. tracking documentation. We should only have one.
    - Copy the contents of the second, smaller page into the first and then delete that content from the second page.
- [x] Mention parallel coordinate and time-series plots.

A note on each checked item:

1. I checked this using Kedro 0.18.3, since 0.18.4 will prevent the charts from loading. There were a couple of minor hangups while running through the tutorial that I've remedied. All should work smoothly now.
2. I've deleted this [secondary docs page](https://kedro.readthedocs.io/en/latest/logging/experiment_tracking.html) with experiment tracking instructions and didn't move anything over into the main page because it was already well covered.
3. The new work is captured and takes precedent in the ordering on the page, meaning the metrics plots in the experiment tracking page shows up earlier in the documentation versus the metrics plot that can be viewed on the flowchart page.

Lastly, I've updated the gifs and images to reflect the changes in design on the experiment tracking page (now we have tabs for overview, metrics, and plots, and before we didn't).

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes


<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/2193"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

tynandebold commented on docs/source/visualisation/experiment_tracking.md:
```diff
@@ -1,4 +1,4 @@
-# Set up experiment tracking
+# Experiment tracking in Kedro-Viz
```
I've updated the wording here to be consistent with the other titles of the sections under `Visualisation with Kedro-Viz`.

tynandebold commented on docs/source/visualisation/experiment_tracking.md:
````diff
@@ -199,18 +197,18 @@ def create_confusion_matrix(companies: pd.DataFrame):
 And now add this node to the `data_processing` pipeline (`src/kedro-experiment-tracking-tutorial/pipelines/data_processing/pipeline.py`)
 
 ```python
+from .nodes import create_confusion_matrix
````
This import was missing. Without it, doing a `kedro run` failed.

stichbury commented on docs/source/visualisation/experiment_tracking.md:
```diff
@@ -8,13 +8,12 @@ Experiment tracking is the process of saving all machine-learning related experi
 
 Enabling experiment tracking features on Kedro-Viz relies on:
```
I'd be tempted to make this more imperative.

```suggestion
To enable the experiment tracking features of Kedro-Viz you need to:
```

stichbury commented on docs/source/visualisation/experiment_tracking.md:
```diff
@@ -8,13 +8,12 @@ Experiment tracking is the process of saving all machine-learning related experi
-* [setting up a session store to capture experiment metadata](#set-up-the-session-store),
-* [experiment tracking datasets to let Kedro know what metrics should be tracked](#set-up-tracking-datasets),
-* [modifying your nodes and pipelines to output those metrics](#set-up-your-nodes-and-pipelines-to-log-metrics).
+- [Setting up a session store to capture experiment metadata](#set-up-the-session-store),
```
```suggestion
- [Set up a session store to capture experiment metadata](#set-up-the-session-store),
```

stichbury commented on docs/source/visualisation/experiment_tracking.md:
```diff
@@ -8,13 +8,12 @@ Experiment tracking is the process of saving all machine-learning related experi
-* [experiment tracking datasets to let Kedro know what metrics should be tracked](#set-up-tracking-datasets),
-* [modifying your nodes and pipelines to output those metrics](#set-up-your-nodes-and-pipelines-to-log-metrics).
+- [Setting up a session store to capture experiment metadata](#set-up-the-session-store),
+- [Experiment tracking datasets to let Kedro know what metrics should be tracked](#set-up-tracking-datasets),
```
```suggestion
- [Set up experiment tracking datasets to list the metrics to track](#set-up-tracking-datasets),
```

stichbury commented on docs/source/visualisation/experiment_tracking.md:
```diff
@@ -8,13 +8,12 @@ Experiment tracking is the process of saving all machine-learning related experi
-* [modifying your nodes and pipelines to output those metrics](#set-up-your-nodes-and-pipelines-to-log-metrics).
+- [Setting up a session store to capture experiment metadata](#set-up-the-session-store),
+- [Experiment tracking datasets to let Kedro know what metrics should be tracked](#set-up-tracking-datasets),
+- [Modifying your nodes and pipelines to output those metrics](#set-up-your-nodes-and-pipelines-to-log-metrics).
```
```suggestion
- [Modify your nodes and pipelines to output those metrics](#set-up-your-nodes-and-pipelines-to-log-metrics).
```

stichbury commented:
This looks great, thanks @tynandebold.

I do have a query about Spaceflights. So this now works if you take the most recent spaceflights starter and add the code given above? There's no reliance on any previous part of the Kedro-Viz docs like the previous page on Plotly reporting? I would normally test it myself just to 'dogfood' the instructions, but I'm a bit tied up with the Academy decks today and don't want to hold you up.

I think we still need to introduce Kedro-Viz for anyone (unlikely enough) to be coming straight to this page without having previously installed Viz. [We put this into the Plotly page.](https://github.com/kedro-org/kedro/blob/main/docs/source/visualisation/visualise_charts_with_plotly.md#visualisation-with-plotly) -- would it be possible to add to this file?

stichbury commented:
This looks great to me @tynandebold. LMK when you've made those changes and I'll approve.

tynandebold commented:
I've made the changes now @stichbury. I took your second suggestion and moved the testing of the Viz version to after the install step. Let me know how that reads and if it's in the correct place.

And to your query about Spaceflights: yes, all worked fine for me when I tested with the latest version of Spaceflights (using the `kedro new` command) and added the code from this PR. I used a fresh conda environment for testing this and followed the instruction from this page only, taking nothing from any other page in the docs. 

Could there have been some leftover dependencies hanging around in my new conda environment that were populated from an older one? I'm not too familiar with this. Again, it worked fine for me with 0.18.3 of Kedro and the couple new lines that I added, so I'm quite confident it's solid. I do hope to test with Merel's fix though, to ensure it'll work with 0.18.4.

stichbury commented on docs/source/visualisation/experiment_tracking.md:
```diff
@@ -1,27 +1,26 @@
-* [setting up a session store to capture experiment metadata](#set-up-the-session-store),
-* [experiment tracking datasets to let Kedro know what metrics should be tracked](#set-up-tracking-datasets),
-* [modifying your nodes and pipelines to output those metrics](#set-up-your-nodes-and-pipelines-to-log-metrics).
+- [Set up a session store to capture experiment metadata](#set-up-the-session-store),
+- [Set up experiment tracking datasets to list the metrics to track](#set-up-tracking-datasets),
+- [Modify your nodes and pipelines to output those metrics](#set-up-your-nodes-and-pipelines-to-log-metrics).
```
```suggestion
- [set up a session store to capture experiment metadata](#set-up-the-session-store),
- [set up experiment tracking datasets to list the metrics to track](#set-up-tracking-datasets),
- [modify your nodes and pipelines to output those metrics](#set-up-your-nodes-and-pipelines-to-log-metrics).
```

stichbury replied:
Sorry, @tynandebold This was my introduction -- I forgot the guidelines for bullets 🤦

tynandebold replied:
Now I'm confused. Should they be capitalized or lowercase?

stichbury replied:
You know what, I'll roll it back to capitals, since other pages of the docs do have them capitalised. I've no idea when/if that's a change, as I may be thinking of another style guide, but that doesn't explain why the bullets in the original copy were lower case here. Never mind. I'm now happy for you to merge this when you are ready.

tynandebold replied:
Copy that!

stichbury reviewed (approved):
LGTM, thanks for this! 🌟 🌟 🌟 🌟 🌟

stichbury commented on docs/source/visualisation/experiment_tracking.md:
```diff
@@ -1,27 +1,26 @@
-* [setting up a session store to capture experiment metadata](#set-up-the-session-store),
-* [experiment tracking datasets to let Kedro know what metrics should be tracked](#set-up-tracking-datasets),
-* [modifying your nodes and pipelines to output those metrics](#set-up-your-nodes-and-pipelines-to-log-metrics).
+- [set up a session store to capture experiment metadata](#set-up-the-session-store),
```
```suggestion
- [Set up a session store to capture experiment metadata](#set-up-the-session-store),
```

stichbury commented on docs/source/visualisation/experiment_tracking.md:
```diff
@@ -1,27 +1,26 @@
-* [experiment tracking datasets to let Kedro know what metrics should be tracked](#set-up-tracking-datasets),
-* [modifying your nodes and pipelines to output those metrics](#set-up-your-nodes-and-pipelines-to-log-metrics).
+- [Set up a session store to capture experiment metadata](#set-up-the-session-store),
+- [set up experiment tracking datasets to list the metrics to track](#set-up-tracking-datasets),
```
```suggestion
- [Set up experiment tracking datasets to list the metrics to track](#set-up-tracking-datasets),
```

stichbury commented on docs/source/visualisation/experiment_tracking.md:
```diff
@@ -1,27 +1,26 @@
-* [modifying your nodes and pipelines to output those metrics](#set-up-your-nodes-and-pipelines-to-log-metrics).
+- [Set up a session store to capture experiment metadata](#set-up-the-session-store),
+- [Set up experiment tracking datasets to list the metrics to track](#set-up-tracking-datasets),
+- [modify your nodes and pipelines to output those metrics](#set-up-your-nodes-and-pipelines-to-log-metrics).
```
```suggestion
- [Modify your nodes and pipelines to output those metrics](#set-up-your-nodes-and-pipelines-to-log-metrics).
```

merelcht reviewed (approved):
This looks good to me! And so exciting to see the time series and parallel coordinates! 🥳

yetudada commented on docs/source/visualisation/experiment_tracking.md:
```diff
@@ -1,37 +1,53 @@
 ![](../meta/images/experiment-tracking_demo_small.gif)
 
-Enabling experiment tracking features on Kedro-Viz relies on:
+## When should I use experiment tracking in Kedro?
```
I will need help from @stichbury and @NeroOkwa in this new section. Does this content work?

yetudada replied:
From @NeroOkwa: "yeah looks great!"

stichbury replied:
Yep, looks good to me. I've made one suggestion but it's not a huge issue so you can leave as you've edited if it's simpler.

yetudada commented on docs/source/visualisation/experiment_tracking.md:
````diff
@@ -43,15 +59,15 @@ SESSION_STORE_CLASS = SQLiteStore
+Set up two datasets to log the columns used in the companies dataset (`companies_columns`) and experiment metrics for the `active_modelling_pipeline` (`active_modelling_pipeline.metrics`) like the coefficient of determination (`r2 score`),  max error (`me`) and mean absolute error (`mae`) by adding the following in the `conf/base/catalog.yml` file:
 
 ```yaml
 active_modelling_pipeline.metrics:
````
My only other comment is based on this and it might need to be a bigger change to how we structure this tutorial. We introduce namespacing here and in the plot comparison section (`data_processing.confusion_matrix` in `catalog.yml`) but we don't actually explain what the full stop means here. 

It's important that we do because I got confused with the guidance in 113 - `Note that the output dataset must exactly match the name of the tracking dataset specified in the catalog file.` because users are asked to just specify `metrics` and not `active_modelling_pipeline.metrics`. 

So my question would be: 
 - Can we, ideally, do this tutorial without introducing the concept of namespaces too (it's an information overload)?
 - And if not, how do we introduce it?

merelcht replied:
Great spot @yetudada ! We do indeed not need the namespace here, as no namespacing has been applied to the pipelines.

yetudada replied:
Awesome let me update the other mentions of `active_modelling_pipeline.metrics` 😊

yetudada commented on docs/source/visualisation/experiment_tracking.md:
````diff
@@ -108,7 +123,7 @@ node(
 ```
 
-Repeat the same steps to set up the `companies_column` dataset. For this dataset, log the column that contains the list of companies as outlined in the `companies.csv` file under the `/data/01_raw` directory. Modify the `preprocess_companies` node under the `data_processing` pipeline (`src/kedro-experiment-tracking-tutorial/pipelines/data_processing/nodes.py`) to return the data under a key value ...
+Repeat the same steps to set up the `companies_column` dataset. For this dataset, log the column that contains the list of companies as outlined in the `companies.csv` file under the `data/01_raw` directory. Modify the `preprocess_companies` node under the `data_processing` pipeline (`src/kedro-experiment-tracking-tutorial/pipelines/data_processing/nodes.py`) to return the data under a key value  ...
````
```suggestion
Repeat the same steps to set up the `companies_column` dataset. For this dataset, log the column that contains the list of companies as outlined in the `companies.csv` file under the `data/01_raw` directory. Modify the `preprocess_companies` node under the `data_processing` pipeline (`src/kedro-experiment-tracking-tutorial/pipelines/data_processing/nodes.py`) to return the data under a key-value pair, as shown below:
```

yetudada reviewed (changes requested):
Thank you so much for this stellar work @tynandebold! I've made some additions and I've left a bigger question which we might need @rashidakanchwala or @merelcht to help with.

yetudada commented on docs/source/visualisation/experiment_tracking.md:
````diff
@@ -43,15 +59,15 @@ SESSION_STORE_CLASS = SQLiteStore
+Set up two datasets to log the columns used in the companies dataset (`companies_columns`) and experiment metrics for the `active_modelling_pipeline` (`active_modelling_pipeline.metrics`) like the coefficient of determination (`r2 score`),  max error (`me`) and mean absolute error (`mae`) by adding the following in the `conf/base/catalog.yml` file:
 
 ```yaml
 active_modelling_pipeline.metrics:
````
```suggestion
metrics:
```

yetudada commented on docs/source/visualisation/experiment_tracking.md:
```diff
@@ -43,15 +59,15 @@ SESSION_STORE_CLASS = SQLiteStore
 There are two types of tracking datasets: [`tracking.MetricsDataSet`](/kedro.extras.datasets.tracking.MetricsDataSet) and [`tracking.JSONDataSet`](/kedro.extras.datasets.tracking.JSONDataSet). The `tracking.MetricsDataSet` should be used for tracking numerical metrics, and the `tracking.JSONDataSet` can be used for tracking any other JSON-compatible data like boolean or text-based data.
 
-Set up two datasets to log `r2 scores` and `parameters` for each run by adding the following in the `conf/base/catalog.yml` file:
+Set up two datasets to log the columns used in the companies dataset (`companies_columns`) and experiment metrics for the `active_modelling_pipeline` (`active_modelling_pipeline.metrics`) like the coefficient of determination (`r2 score`),  max error (`me`) and mean absolute error (`mae`) by adding the following in the `conf/base/catalog.yml` file:
```
```suggestion
Set up two datasets to log the columns used in the companies dataset (`companies_columns`) and experiment metrics for the data science pipeline (`metrics`) like the coefficient of determination (`r2 score`),  max error (`me`) and mean absolute error (`mae`) by adding the following in the `conf/base/catalog.yml` file:
```

stichbury commented on docs/source/visualisation/experiment_tracking.md:
```diff
@@ -1,37 +1,53 @@
 
-This page describes the steps necessary to set up experiment tracking and access logged metrics, using the [spaceflights tutorial](../tutorial/spaceflights_tutorial.md).
+ - **Kedro** - If you need experiment tracking, are looking for improved metrics visualisation and want a lightweight tool created to leverage existing functionality in Kedro. Kedro does not support a model registry. 
+ - **MLflow** - If you require experiment tracking, model registry and/or model serving capabilities or have access to Managed MLflow within the Databricks ecosystem. Experiment tracking in Kedro is a visual upgrade of your experience using MLFlow Tracking. Leverage [`kedro-mlflow`](https://kedro-mlflow.readthedocs.io/en/stable/) for your workflow. 
```
I have a bit of a problem with "Experiment tracking in Kedro is a visual upgrade of your experience using MLFlow Tracking. Leverage `kedro-mlflow` for your workflow. " as it sounds like marketing and doesn't explain how Kedro upgrades MLFlow Tracking visually.

Can I suggest:

```suggestion
 - **MLflow** - You can combine MLFlow with Kedro by using [`kedro-mlflow`](https://kedro-mlflow.readthedocs.io/en/stable/) if you require experiment tracking, model registry and/or model serving capabilities or have access to Managed MLflow within the Databricks ecosystem.  
```

stichbury commented:
> I think we just need a general check to see if this tutorial runs from beginning to end in a clean virtual environment with all the changes but otherwise I think we're ready to go live.

I'm happy to try this out as I want to test it out and learn about experiment tracking, but I'm not really able to right now. I think @tynandebold needs to hold off until the next viz release anyway, so hopefully this can wait until next week -- I'll get it done before the end of the sprint.

tynandebold commented:
@stichbury you'll have a couple days to test it out. Hopefully we release Viz tomorrow or Wednesday.

Separately, in this PR I think we should remove the references to Kedro version `0.18.3` [in this file](https://github.com/kedro-org/kedro/blob/main/docs/source/visualisation/visualise_charts_with_plotly.md), since that problem should be solved when this [PR is completed](https://github.com/kedro-org/kedro-viz/pull/1214). What do you think?

stichbury commented on docs/source/visualisation/experiment_tracking.md:
````diff
@@ -173,9 +188,25 @@ You can now access, compare and pin your runs by toggling the `Compare runs` but
+You must update the `src/requirements.txt` file in your Kedro project by adding the following dataset to enable Matplotlib for your project:
+
+```bash
+kedro[matplotlib.MatplotlibWriter]==0.18.3
````
True! I have to make some wholesale changes to the other viz docs when that PR merges so can remove this then if you merge this PR ahead of #1214 but otherwise, if it won't merge in until later, you can certainly remove this.

tynandebold replied:
I've removed it since this PR will be merged in the next couple of days.

stichbury commented:
> @stichbury you'll have a couple days to test it out. Hopefully we release Viz tomorrow or Wednesday.
> 
> Separately, in this PR I think we should remove the references to Kedro version `0.18.3` [in this file](https://github.com/kedro-org/kedro/blob/main/docs/source/visualisation/visualise_charts_with_plotly.md), since that problem should be solved when this [PR is completed](https://github.com/kedro-org/kedro-viz/pull/1214). What do you think?

I'll prioritise this on Tuesday 17th.

stichbury commented:
> > @stichbury you'll have a couple days to test it out. Hopefully we release Viz tomorrow or Wednesday.
> > Separately, in this PR I think we should remove the references to Kedro version `0.18.3` [in this file](https://github.com/kedro-org/kedro/blob/main/docs/source/visualisation/visualise_charts_with_plotly.md), since that problem should be solved when this [PR is completed](https://github.com/kedro-org/kedro-viz/pull/1214). What do you think?
> 
> I'll prioritise this on Tuesday 17th.

@tynandebold I've now tested the text and all is good. I made a couple of changes to the experiment tracking page to tweak it (nothing major) and also removed mention of Kedro 0.18.3 in the Viz docs about plotly charts, which I've sneaked into this PR. Hope that's OK!

tynandebold commented:
Fantastic, @stichbury! Thank you 💥
